File: php/bowling/bowling.php
<?php
// YAY ! all test passed, take this Uncle Bob
class Game{
    # rolls history
    public $rolls = [];
    public $cur_roll = 0;

    # frames history
    public $frames = [];
    public $cur_frame = 0;

    function __construct(){
        // start with a new frame
        $this->new_frame();
    }

    // record a new frame
    function new_frame(){
        $this->cur_frame++;
        $frame = false;
        if($this->cur_frame < 10){
            $frame = new Frame($this);
        }elseif($this->cur_frame == 10){  # we need a specific set of rules for the 10th frame
            $frame = new TenthFrame($this);
        }
        if($frame){
            $this->frames[$this->cur_frame] = $frame;
        }
    }

    // record a new roll
    function roll(int $pins){
        if($pins < 0 || $pins > 10){
            throw new Exception("0 <= pins <= 10");
        }
        // record rolls in common history
        $this->cur_roll++;
        $this->rolls[$this->cur_roll] = $pins;
        // record roll in the current frame
        $this->frames[$this->cur_frame]->roll($pins);
    }

    // sum of all frame scores
    function score(): int{
        $total = 0;
        foreach($this->frames as $k => $frame){
            $total += $frame->score();
        }
        return $total;
    }
}


/*
maybe something like trait would be more adequate for designing this,
but i don't master this yet
*/
class CommonFrame{
    public $pins = 10;
    public $roll = 2;
    public $score = [];
    public $type = "UNFINISHED";

    protected $_parent;

    function __construct($parent){
        $this->_parent = $parent; # so we could access to the common game data from the frame
    }

    // record roll in frame
    function roll(int $pins){
        $this->pins -= $pins;
        if($this->pins < 0){
            throw new Exception("someone is cheating ?");
        }
        $this->roll --;
        $this->score[$this->_parent->cur_roll] = $pins;
    }

    // current frame score
    function score(){
        if($this->type === "UNFINISHED"){
            throw new Exception("UNFINISHED"); # the unit tests don't want us to calculate an unfinished game
        }
    }
}


class Frame extends CommonFrame{
    function roll($pins){
        parent::roll($pins);

        if(!$this->roll || !$this->pins){
            if(!$this->pins && $this->roll){
                $this->type = "STRIKE";
            }elseif(!$this->pins && !$this->roll){
                $this->type = "SPARE";
            }else{
                $this->type = "OPEN";
            }

            $this->_parent->new_frame();
        }
    }

    function score(){
        parent::score();
        $key = key($this->score);

        # look ahead
        switch($this->type){
            case "STRIKE":
            case "SPARE":
                $length = 3;
                break;
            case "OPEN":
                $length = 2;
                break;
            default:
                $length = 0;
        }
        $slice = array_slice($this->_parent->rolls, $key-1, $length);

        return array_sum($slice);

    }
}


class TenthFrame extends CommonFrame{
    function __construct($parent){
        $this->bonus = 0;
        parent::__construct($parent);
    }

    function roll($pins){
        parent::roll($pins);

        // new frame if roll >=2 or no more pins
        if(!$this->roll || !$this->pins){
            if(!$this->pins && $this->roll){
                $this->type = "STRIKE";
                $this->pins = 10;
                $this->bonus = 1;
            }elseif(!$this->pins && !$this->roll){
                $this->type = "SPARE";
                $this->pins = 10;
                $this->bonus = 1;
            }else{
                $this->type = "OPEN";
            }
        }

        if($this->roll < 0 && !$this->bonus){
            throw new Exception("too much");
        }
    }

    function score(){
        parent::score();
        if($this->bonus && $this->roll != -1){
            throw new Exception("not enough");
        }
        if(!$this->bonus && count($this->score) > 2){
            array_pop($this->score);
        }
        $total =  array_sum($this->score);
        return $total;
    }
}

File: python/bowling/bowling.py
class BowlingGame:
    def __init__(self):
        self._rolls = []
        self._frames = []
        self.new_frame()

    @property
    def _frame_cursor(self):
        return len(self._frames) + 1

    def new_frame(self):
        """ record a new frame """

        new_frame = False

        if self._frame_cursor < 10:
            new_frame = Frame(self)
        elif self._frame_cursor == 10:
            new_frame = TenthFrame(self)

        if new_frame != False:
            self._frames.append(new_frame)

    def roll(self, pins):
        """ Record a new roll """

        # validate pins value
        if not (0 <= pins <= 10):
            raise ValueError("0 <= pins <= 10")

        # create a new Roll
        new_roll = Roll(pins)
        # record it in common history
        self._rolls.append(new_roll)
        # and in the current frame
        self._frames[-1].frame_roll(new_roll)

    def score(self):
        """ Total score """
        return sum(frame.frame_score() for frame in self._frames)


class Roll:
    """ A roll """

    def __init__(self, pins):
        self.pins = pins

    def __repr__(self):
        return "{}".format(self.pins)


class CommonFrame:
    """ Common frame mechanics """

    def __init__(self, parent):
        self._pins = 10
        self._roll = 2
        self._score = []
        self._parent = parent

    @property
    def _is_finished(self):
        return self._pins == 0 or self._roll == 0

    @property
    def _type(self):
        empty_roll = self._roll == 0
        empty_pins = self._pins == 0
        return {
            (True, False): "STRIKE",
            (True, True): "SPARE",
            (False, True): "OPEN",
            (False, False): "UNFINISHED",
        }[(empty_pins, empty_roll)]

    def frame_roll(self, roll):
        """ record a new Roll

        Arguments:
            roll {Roll} -- a Roll object

        Raises:
            ValueError -- impossible pins value
        """

        self._pins -= roll.pins
        if self._pins < 0:
            raise ValueError("is someone cheating ?")
        self._roll -= 1
        self._score.append(roll)


class Frame(CommonFrame):
    """ Normal frame """

    def frame_roll(self, roll):
        super().frame_roll(roll)

        if self._is_finished:
            self._parent.new_frame()

    def frame_score(self):
        """ Calculate frame score
        (with look ahead)

        Returns:
            int -- total frame score
        """

        # look ahead
        length = {
            "STRIKE": 3,
            "SPARE": 3,
            "OPEN": 2,
            "UNFINISHED": 0
        }[self._type]

        key = self._parent._rolls.index(self._score[0])
        rolls = self._parent._rolls[key:key+length]
        return sum(roll.pins for roll in rolls)


class TenthFrame(CommonFrame):
    """ Special frame """

    def __init__(self, parent):
        self._bonus = 0
        super().__init__(parent)

    def frame_roll(self, roll):
        super().frame_roll(roll)

        if self._is_finished:
            # bonus roll
            if self._type in ["STRIKE", "SPARE"]:
                self._pins = 10
                self._bonus = 1

        if (self._roll < 0 and not self._bonus) or self._roll < -1:
            raise IndexError("Cannot roll another one")

    def frame_score(self):
        """ Calculate frame score

        Raises:
            IndexError -- If the frame is unfinished

        Returns:
            int -- frame total score
        """

        if self._bonus and self._roll != -1:
            raise IndexError("Unfinished game")

        return sum(roll.pins for roll in self._score)
